azurelinuxagent/ga/monitor.py

```diff
@@ -424,14 +424,11 @@ def send_telemetry_heartbeat(self):
                 dropped_packets = self.osutil.get_firewall_dropped_packets(self.protocol.get_endpoint())
                 msg = "{0};{1};{2};{3}".format(incarnation, self.counter, self.heartbeat_id, dropped_packets)
 
-                add_event(
-                    name=AGENT_NAME,
-                    version=CURRENT_VERSION,
-                    op=WALAEventOperation.HeartBeat,
-                    is_success=True,
-                    message=msg,
-                    log_event=False)
+                add_event(name=AGENT_NAME, version=CURRENT_VERSION, op=WALAEventOperation.HeartBeat, is_success=True,
+                          message=msg, log_event=False)
```

Comment:
Just a thought, why not just make the `log_event=True` for the `add_event` instead of logging it separately? We only invoke this call once every 30mins so we wouldn't have to worry about making it periodic either.

Reply:
Ohh I see you wanted to log every half_day instead of 30mins. But after the discussion yesterday, you could just log the heartbeat within the add_event function directly.

Reply:
I moved the logging to the extHandler now. Monitor Handler will still send the telemetry and extHandler (being the main thread), will log correctly.

Thus, the above change won't help.

Also on a side note - `log_event=True` has some weird behavior - It only logs when `is_success=False`, and thus, I didn't want to rely on that.

Reply:
I feel we should keep both (telemetry and the log) in the same thread if its meant for the same purpose. I dont like the discrepancy between sending the same data from 2 different points.

Consider this scenario - 
Say ExtHandler was down (which we verify from logs) but we still see heartbeat telemetry in Kusto. The oncall engineer would think that the agent is working fine when in all reality it isint, it'll just confuse the oncall engineer.

Reply:
@larohra - we've already hit that scenario with the current code: we have hearbeat and no activity in the local log.

move the event here?

Reply:
after all, it is the main thread who reports status to CRP so the event/log may be more appropriate here.
the main thread also monitors the other 2 thread to check that they are active.

Reply:
@narrieta I think we're talking about the same thing. I'm also advocating for moving the Telemetry event + log to the main thread rather than doing it in monitor thread. 

(We're commenting in the monitor thread right now)

Reply:
@larohra - I agree with the scenario - and ideally, the heartbeat should come from the main extension processing thread only, but we can't overload the single process to do all the scenarios. This is, IMO, the next best thing we can get, to showcase if the main exthandler is running or not.

Reply:
@vrdmr the monitor thread is the one who's going to actually send out the events to wireserver in the end, its just about who's going to add the heartbeat event file in the filesystem (which is not an expensive operation) so I dont think we're going to be overriding the main thread in any way. Additionally since we're going to be sending out heartbeats every 30mins its not going to be very expensive for the ExtHandler thread to write that file  :)

Reply:
@larohra can you clarify what you mean by "main" thread? The thread that starts other threads and checks if they're alive is the `update.py`, not `exthandlers.py`. You seem to be conflating the two.

I think the `update.py` thread would be appropriate for this kind of logging, since it's the most expensive thread to die and we should care if that one is reporting heartbeat or not.

Reply:
@larohra 

> @narrieta I think we're talking about the same thing. I'm also advocating for moving the Telemetry event + log to the main thread rather than doing it in monitor thread.

agreed

Reply:
also, if we do that I would rather check the time interval to log using the same logic as we do in the other threads rather than relying in the [PERIODIC] logging, since that can be reset (and has been reset in the past) by other components

Reply:
>The thread that starts other threads and checks if they're alive is the update.py, not exthandlers.py

@pgombar they are the same threads, you might be getting confused with the name `UpdateHandler` and `ExtHandler`. Those are the classes that are running in the same `ExtHandler` thead :)

```python
    def run_exthandlers(self, debug=False):
        """
        Run the update and extension handler
        """
        logger.set_prefix("ExtHandler")
        threading.current_thread().setName("ExtHandler")
        from azurelinuxagent.ga.update import get_update_handler
        update_handler = get_update_handler()
        update_handler.run(debug)
```

Reply:
You're totally right, I just wasn't paying attention. Thanks for the correction!

Reply:
@larohra: Agreed. I liked the idea and thus moved the whole HB logic into the UpdateHandler.